### PR TITLE
feat: woodpecker manager

### PR DESCRIPTION
**Pull request by MarcelCoding**
## Changes

Added support for woodpecker pipeline configurations.

## Context

Every step in a pipeline uses a docker image, where the step is being executed. -> Automatically update these images.
I used the docker-compose manager as a starting point.

Json Schema:
https://raw.githubusercontent.com/woodpecker-ci/woodpecker/master/pipeline/schema/schema.json

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

Can someone verify I changed all the necessary things to add a new package manager? I don't really have any idea about this codebase.

### CODE REVIEW

File: lib/modules/manager/woodpecker/readme.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Extracts all Docker images from with Woodpecker Pipeline YAML files.
```
Comment (HonkingGoose) [inline lib/modules/manager/woodpecker/readme.md:1]:
```suggestion
Extracts all Docker images from Woodpecker Pipeline YAML files.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/config/presets/internal/docker.ts [hunk 2]
```diff
@@ -12,6 +12,9 @@ export const presets: Record<string, Preset> = {
     circleci: {
       enabled: false,
     },
+    woodpecker: {
+      enabled: true,
+    },
```
Comment (viceice) [inline lib/config/presets/internal/docker.ts:17]: remove, unrelated

File: lib/modules/manager/woodpecker/extract.spec.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      expect(extractPackageFile('nothing here', '', {})).toBeNull();
+    });
+
+    it('returns null for malformed YAML', () => {
+      expect(extractPackageFile('nothing here\n:::::::', '', {})).toBeNull();
+    });
+
+    it('extracts multiple image lines for version 1', () => {
+      const res = extractPackageFile(yamlFile, '', {});
+      expect(res?.deps).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/manager/woodpecker/extract.spec.ts:22]: use toMatchObject 

read our best practices guide, checkout docs/development folder

File: lib/modules/manager/woodpecker/extract.spec.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      expect(res?.deps).toHaveLength(8);
+    });
+
+    it('extracts multiple image lines for version 3', () => {
+      const res = extractPackageFile(yamlFile, '', {});
+      expect(res?.deps).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/manager/woodpecker/extract.spec.ts:28]: same here

File: lib/modules/manager/woodpecker/extract.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  try {
+    const lineMapper = new LineMapper(content, regEx(/^\s*image:/));
+
+    // Image name/tags for services are only eligible for update if they don't
+    // use variables and if the image is not built locally
+    const deps = Object.values(config.pipeline ?? {})
+      .filter((step) => is.string(step?.image))
+      .map((step) => {
+        const dep = getDep(step.image, true, extractConfig.registryAliases);
+        const lineNumber = lineMapper.pluckLineNumber(step.image);
```
Comment (viceice) [inline lib/modules/manager/woodpecker/extract.ts:69]: remove line number handling, not used and not required

File: lib/modules/manager/woodpecker/extract.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+          return null;
+        }
+        return dep;
+      })
+      .filter(is.truthy);
+
+    logger.trace({ deps }, 'Woodpecker Configuration image');
+    return { deps };
+  } catch (err) /* istanbul ignore next */ {
+    logger.warn(
```
Comment (viceice) [inline lib/modules/manager/woodpecker/extract.ts:81]:
```suggestion
    logger.debug(
```

File: lib/modules/manager/woodpecker/extract.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        // istanbul ignore if
+        if (!lineNumber) {
+          return null;
+        }
+        return dep;
+      })
+      .filter(is.truthy);
+
+    logger.trace({ deps }, 'Woodpecker Configuration image');
+    return { deps };
```
Comment (viceice) [inline lib/modules/manager/woodpecker/extract.ts:79]: return null if deps is empty

File: lib/modules/manager/woodpecker/extract.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    if (typeof config !== 'object') {
+      logger.debug(
+        { fileName, type: typeof config },
+        'Unexpected type for Woodpecker Configuration content'
+      );
+      return null;
+    }
+  } catch (err) {
+    logger.debug({ err }, 'err');
+    logger.debug({ fileName }, 'Parsing Woodpecker Configuration config YAML');
```
Comment (viceice) [inline lib/modules/manager/woodpecker/extract.ts:57]: combine to single log entry

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/woodpecker/extract.ts (hunk 7 above)
Comment (MarcelCoding) [reply lib/modules/manager/woodpecker/extract.ts:79]: The object or the array?
`null` or `{deps: null}`

**Comment by MarcelCoding** [comment]
@viceice I've done the tests a bit differently. The toMatchedObject method only works well on objects, this is an array. I hope my method is also ok.

File: lib/modules/manager/docker-compose/extract.spec.ts [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import { Fixtures } from '../../../../test/fixtures';
 import { extractPackageFile } from '.';
 
-const yamlFile1 = Fixtures.get('docker-compose.1.yml');
+const yamlFile1 = Fixtures.get('.woodpecker.yml');
```
Comment (viceice) [inline lib/modules/manager/docker-compose/extract.spec.ts:4]: this seems wrong to me, this PR shouldn't change docker-compose manager

File: lib/modules/manager/woodpecker/extract.spec.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
+import { Fixtures } from '../../../../test/fixtures';
+import * as expectedDeps from './__fixtures__/deps.json';
```
Comment (viceice) [inline lib/modules/manager/woodpecker/extract.spec.ts:2]: inline those

File: lib/modules/manager/woodpecker/extract.spec.ts [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    });
+
+    it('returns null for malformed YAML', () => {
+      expect(extractPackageFile('nothing here\n:::::::', '', {})).toBeNull();
+    });
+
+    it('extracts multiple image lines', () => {
+      const res = extractPackageFile(yamlFile, '', {});
+      expect(res?.deps).toEqual(expectedDeps.deps);
+      expect(res?.deps).toHaveLength(8);
```
Comment (viceice) [inline lib/modules/manager/woodpecker/extract.spec.ts:24]: remove

File: lib/modules/manager/woodpecker/index.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import { ProgrammingLanguage } from '../../../constants';
+import { DockerDatasource } from '../../datasource/docker';
+import { extractPackageFile } from './extract';
+
+export const language = ProgrammingLanguage.Docker;
+
+export { extractPackageFile };
+
+export const defaultConfig = {
+  fileMatch: ['(^|/)\\.woodpecker[^/]*\\.ya?ml$'],
```
Comment (viceice) [inline lib/modules/manager/woodpecker/index.ts:10]:
```suggestion
  fileMatch: ['(^|\\/)\\.woodpecker[^/]*\\.ya?ml$'],
```
properly escape

**Comment by viceice** [comment]
> @viceice I've done the tests a bit differently. The toMatchedObject method only works well on objects, this is an array. I hope my method is also ok.

that works also well for arrays, it needs do be same length. but toEqual is even better. you should validate the whole response, not only the deps array.

**Comment by MarcelCoding** [comment]
@viceice sorry, IntelliJ file named did the change in docker-compose, I fixed it :)

File: lib/modules/manager/woodpecker/extract.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    // Image name/tags for services are only eligible for update if they don't
+    // use variables and if the image is not built locally
+    const deps = Object.values(config.pipeline ?? {})
+      .filter((step) => is.string(step?.image))
+      .map((step) => getDep(step.image, true, extractConfig.registryAliases))
+      .filter(is.truthy);
+
+    logger.trace({ deps }, 'Woodpecker Configuration image');
+    return deps.length ? { deps } : null;
+  } catch (err) {
```
Comment (viceice) [inline lib/modules/manager/woodpecker/extract.ts:49]: when we got to this catch? there seems no code which ever should throw

File: lib/modules/manager/woodpecker/index.ts [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import { ProgrammingLanguage } from '../../../constants';
+import { DockerDatasource } from '../../datasource/docker';
+import { extractPackageFile } from './extract';
+
+export const language = ProgrammingLanguage.Docker;
```
Comment (viceice) [inline lib/modules/manager/woodpecker/index.ts:5]: does woodpecker support other kinds of dependencies beside docker images? are there other types planned?

File: lib/modules/manager/woodpecker/readme.md [hunk 15]
```diff
@@ -0,0 +1,3 @@
+Extracts all Docker images from Woodpecker Pipeline YAML files.
```
Comment (viceice) [inline lib/modules/manager/woodpecker/readme.md:1]: please add se links to woodpecker homepage / documentation

File: lib/modules/manager/woodpecker/index.ts (hunk 14 above)
Comment (MarcelCoding) [reply lib/modules/manager/woodpecker/index.ts:5]: No, currently only Docker Images are supproted. Another thing would be to use `apt`, `apk` or `dnf` ... in a comment with a pinned package version, but I think that is out of this project's scope and not really doable.

**Comment by MarcelCoding** [comment]
I added links to the documentation and removed the unnecessary try-catch.

File: lib/modules/manager/woodpecker/readme.md [hunk 16]
```diff
@@ -0,0 +1,9 @@
...
+
+- [Homepage](https://woodpecker-ci.org/)
```
Comment (HonkingGoose) [inline lib/modules/manager/woodpecker/readme.md:3]:
```suggestion
- [Woodpecker homepage](https://woodpecker-ci.org/)
```

File: lib/modules/manager/woodpecker/readme.md [hunk 17]
```diff
@@ -0,0 +1,9 @@
...
+- [Pipeline Syntax Documentation](https://woodpecker-ci.org/docs/usage/pipeline-syntax)
+- [Json Schema](https://raw.githubusercontent.com/woodpecker-ci/woodpecker/master/pipeline/schema/schema.json)
```
Comment (HonkingGoose) [inline lib/modules/manager/woodpecker/readme.md:5]:
```suggestion
- [`woodpecker-ci` JSON schema](https://raw.githubusercontent.com/woodpecker-ci/woodpecker/master/pipeline/schema/schema.json)
```

File: lib/modules/manager/woodpecker/readme.md [hunk 18]
```diff
@@ -0,0 +1,9 @@
+Extracts all Docker images from Woodpecker Pipeline YAML files.
+
+- [Homepage](https://woodpecker-ci.org/)
+- [Pipeline Syntax Documentation](https://woodpecker-ci.org/docs/usage/pipeline-syntax)
```
Comment (HonkingGoose) [inline lib/modules/manager/woodpecker/readme.md:4]:
```suggestion
- [Woodpecker Docs: Pipeline Syntax](https://woodpecker-ci.org/docs/usage/pipeline-syntax)
```

File: lib/modules/manager/woodpecker/readme.md [hunk 19]
```diff
@@ -0,0 +1,9 @@
...
+- [Json Schema](https://raw.githubusercontent.com/woodpecker-ci/woodpecker/master/pipeline/schema/schema.json)
+
+- [Section where Dependencies in yaml files are being used [Documentation]](https://woodpecker-ci.org/docs/usage/pipeline-syntax#image)
```
Comment (HonkingGoose) [inline lib/modules/manager/woodpecker/readme.md:7]:
```suggestion
- [Section where Dependencies in YAML files are being used [Documentation]](https://woodpecker-ci.org/docs/usage/pipeline-syntax#image)
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
The link names should be clear about where you'll end up after you click. 😉

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/woodpecker/readme.md [hunk 20]
```diff
@@ -0,0 +1,9 @@
+Extracts all Docker images from Woodpecker Pipeline YAML files.
+
+- [Woodpecker homepage](https://woodpecker-ci.org/)
+- [Woodpecker Docs: Pipeline Syntax](https://woodpecker-ci.org/docs/usage/pipeline-syntax)
+- [`woodpecker-ci` JSON schema](https://raw.githubusercontent.com/woodpecker-ci/woodpecker/master/pipeline/schema/schema.json)
+
+- [Section where Dependencies in YAML files are being used [Documentation]](https://woodpecker-ci.org/docs/usage/pipeline-syntax#image)
```
Comment (HonkingGoose) [inline lib/modules/manager/woodpecker/readme.md:7]: We have two links that go to `docs/usage/pipeline-syntax` but one link has the `#image` fragment in the link. Maybe we can use just the first pipeline docs link and explain that the reader should also check the section about "Where dependencies in YAML files are being used"?

Our docs feel like: "Have some links, you'll figure it out.". Is this what we intend for the Woodpecker manager docs? Do we want some examples in our docs maybe?

Comment (MarcelCoding) [reply lib/modules/manager/woodpecker/readme.md]: How about the new version?

**Review by HonkingGoose — DISMISSED** [review]

File: lib/modules/manager/woodpecker/readme.md (hunk 20 above)
Comment (HonkingGoose) [reply lib/modules/manager/woodpecker/readme.md]: I like it! 😉

**Comment by MarcelCoding** [comment]
One question, thouth the signing of the cla, will my email somehow be published? If yes, I would like to use a different one, but with the GitHub OAuth, I couldn't select which one.

**Comment by viceice** [comment]
> One question, thouth the signing of the cla, will my email somehow be published? If yes, I would like to use a different one, but with the GitHub OAuth, I couldn't select which one.

nope, your GitHub account will be marked as cla signed, email address for git commits doesn't matter, it only needs to be assigned to that GitHub account.

File: lib/modules/manager/woodpecker/extract.ts [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+    );
+    return null;
+  }
+
+  // Image name/tags for services are only eligible for update if they don't
+  // use variables and if the image is not built locally
+  const deps = Object.values(config.pipeline ?? {})
+    .filter((step) => is.string(step?.image))
+    .map((step) => getDep(step.image, true, extractConfig.registryAliases))
+    .filter(is.truthy);
```
Comment (viceice) [inline lib/modules/manager/woodpecker/extract.ts:45]: i don't think getDep returns undefined or null, so this check can probably removed

File: lib/modules/manager/woodpecker/extract.ts [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+      logger.debug(
+        { fileName, type: typeof config },
+        'Unexpected type for Woodpecker Configuration content'
+      );
+      return null;
+    }
+  } catch (err) {
+    logger.debug(
+      { fileName, err },
+      'Parsing Woodpecker Configuration config YAML'
```
Comment (rarkins) [inline lib/modules/manager/woodpecker/extract.ts:35]:
```suggestion
      'Error parsing Woodpecker Configuration config YAML'
```

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**